test/parallel/test-http-client-finished.js

```diff
@@ -0,0 +1,110 @@
+'use strict';
+const common = require('../common');
+const http = require('http');
+const { finished } = require('stream')
+
+{
+  // abort before finished
+
+  const server = http.createServer(function(req, res) {
+  });
+
+  server.listen(0, common.mustCall(function() {
+    const req = http.request({
+      port: this.address().port
+    }, common.mustNotCall());
+    req.abort();
+    finished(req, common.mustCall(() => {
+      server.close();
+    }));
+  }));
+}
+
+{
+  // abort after request
+
+  const server = http.createServer(function(req, res) {
+  });
+
+  server.listen(0, common.mustCall(function() {
+    const req = http.request({
+      port: this.address().port
+    }).end();
+    finished(req, (err) => {
+      common.expectsError({
+        type: Error,
+        code: 'ERR_STREAM_PREMATURE_CLOSE'
+      })(err);
+      finished(req, common.mustCall(() => {
+        server.close();
+      }));
+    });
+    req.abort();
+  }));
+}
+
+{
+  // abort before end
+
+  const server = http.createServer(function(req, res) {
+    res.write('test');
+  });
+
+  server.listen(0, common.mustCall(function() {
+    const req = http.request({
+      port: this.address().port
+    }).on('response', common.mustCall(res => {
+      req.abort();
+      finished(res, common.mustCall(() => {
+        finished(res, common.mustCall(() =>{
+          server.close();
+        }));
+      }));
+    })).end();
+  }));
+}
+
+{
+  // destroy before end
+
+  const server = http.createServer(function(req, res) {
+    res.write('test');
+  });
+
+  server.listen(0, common.mustCall(function() {
+    http.request({
+      port: this.address().port
+    }).on('response', common.mustCall(res => {
+      // TODO(ronag): Bug? Won't emit 'close' unless read.
+      res.on('data', () => {});
```

Comment:
@addaleax: Isn't it a bit strange that the `'data'` handler is required here in order for the response to end/close? Likewise in a test further down.

Reply:
ping @addaleax? I think this is out of the scope of this PR. Please just confirm whether this is strange and we can create a new issue for later investigation.